internal/service/auth/auth.go

```diff
@@ -0,0 +1,229 @@
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/google/uuid"
+	db2 "github.com/kamencov/go-musthave-diploma-tpl/internal/storage/db"
+	"time"
+
+	"github.com/kamencov/go-musthave-diploma-tpl/internal/customerrors"
+	"github.com/kamencov/go-musthave-diploma-tpl/internal/service"
+	"github.com/kamencov/go-musthave-diploma-tpl/internal/service/auth/entity"
+)
+
+//go:generate mockgen -source=./auth.go -destination=/Users/pavel/GolandProjects/go-musthave-diploma-tpl/internal/mocks/mock_storage_auth.go -package=mocks
+type AuthService interface {
+	RegisterUser(ctx context.Context, login, password string) error
+	AuthUser(ctx context.Context, login, password string) (entity.Tokens, error)
+	VerifyUser(token string) (string, error)
+	RefreshToken(ctx context.Context, token string) (entity.Tokens, error)
+	GeneratedTokens(ctx context.Context, login string) (entity.Tokens, error)
+	HashPassword(password string) string
+}
+
+type ServiceAuth struct {
+	tokenSalt    []byte
+	passwordSalt []byte
+
+	accessTokenTTL  time.Duration
+	refreshTokenTTL time.Duration
+
+	storage service.Storage
+}
+
+func NewServiceAuth(saltTKN, saltPSW []byte, storage service.Storage) *ServiceAuth {
+	return &ServiceAuth{
+		tokenSalt:    saltTKN,
+		passwordSalt: saltPSW,
+
+		accessTokenTTL:  24 * time.Hour,
+		refreshTokenTTL: 24 * time.Hour,
+
+		storage: storage,
+	}
+}
+
+func (s *ServiceAuth) RegisterUser(ctx context.Context, login, password string) error {
+	err := s.storage.CheckTableUserLogin(ctx, login)
+
+	if errors.Is(err, customerrors.ErrUserAlreadyExists) {
+		return customerrors.ErrUserAlreadyExists
+	}
+
+	if err != nil {
+		return err
+	}
+
+	passwordHash := s.HashPassword(password)
+
+	err = s.storage.SaveTableUser(login, passwordHash)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *ServiceAuth) AuthUser(ctx context.Context, login, password string) (entity.Tokens, error) {
+	passwordHash, ok := s.storage.CheckTableUserPassword(ctx, login)
+	if !ok {
+		return entity.Tokens{}, customerrors.ErrNotFound
+	}
+
+	isPasswordTrue := s.doPasswordMatch(passwordHash, password)
+	if !isPasswordTrue {
+		return entity.Tokens{}, customerrors.ErrIsTruePassword
+	}
+
+	tokens, err := s.GeneratedTokens(ctx, login)
+	if err != nil {
+		return tokens, err
+	}
+
+	return tokens, nil
+}
+
+func (s *ServiceAuth) VerifyUser(token string) (string, error) {
+	claims := &entity.AccessTokenClaims{}
+	parsedToken, err := jwt.ParseWithClaims(token, claims, func(token *jwt.Token) (interface{}, error) {
+		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+			return nil, customerrors.ErrInCorrectMethod
+		}
+
+		return s.tokenSalt, nil
+	})
+
+	if err != nil {
+		return "", fmt.Errorf("incorrect token: %v", err)
+	}
+	if !parsedToken.Valid {
+		return "", fmt.Errorf("token is not valid")
+	}
+
+	return claims.Login, nil
+}
+
+func (s *ServiceAuth) RefreshToken(ctx context.Context, token string) (entity.Tokens, error) {
```

Comment:
Советую использовать указатель на структуру, как возвращаемый тип. Это позволит использовать `nil` в качестве отсутствующего значения.
```suggestion
func (s *ServiceAuth) RefreshToken(ctx context.Context, token string) (*entity.Tokens, error) {
```